recorder/src/main/cpp/perf_ctx.hh

```diff
@@ -0,0 +1,156 @@
+#include <cstdint>
+#include <stack>
+#include <stdexcept>
+#include <array>
+#include <vector>
+#include "util.hh"
+#include <mutex>
+#include <cuckoohash_map.hh>
+#include <city_hasher.hh>
+#include <concurrentqueue.h>
+#include "prob_pct.hh"
+#include <unordered_map>
+#include <functional>
+
+#ifndef PERF_CTX_H
+#define PERF_CTX_H
+
+namespace PerfCtx {
+    typedef std::uint64_t TracePt;
+    /*
+      Ctx bit layout: it is composed of 3 parts, coverage %, merge-semantic and ctx-id and merge-semantic is not relevant for merge-generated ctx
+      because they can never get pushed over some other context (other contexts may get pushed over them, but then the parent's merge-semantic
+      doesn't matter).
+          
+      | bits [0 - 63] = 64 | [63 - 63] = 1 | [62 - 56] = 7 | [55 - 53] = 3  | [52 - 13] = 40 | [12 - 0] = 13                   |
+      |--------------------+---------------+---------------+----------------+----------------+---------------------------------|
+      | user-created       |             0 | coverage %    | merge-semantic | **unused**     | ctx-id (always a prime number)  |
+      |--------------------+---------------+---------------+----------------+----------------+---------------------------------|
+
+      13 bits can house 100th prime number (7919, which is less than 8192), so definitely sufficient for forseeable future.
+
+      | bits [0 - 63] = 64 | [63 - 63] = 1 | [62 - 7] = 56               | [6 - 0] = 7                 |
+      |--------------------+---------------+-----------------------------+-----------------------------|
+      | merge-generated    |             1 | ctx-id part-0 (combination) | ctx-id part-1 (permutation) |
```

Comment:
Couldn't think of a better way to pack it in 64 bits (avoid any table-lookup whatsoever in the fast path, basically signal-handling).